test(kt-devnet): add batcher failover kurtosis test

samlaf opened the pull request:
Closes DAINT-303

This PR adds a golang failover test which uses the kurtosis devnet as backend.
This test is used to test the new batcher failover behavior from https://github.com/Layr-Labs/optimism/pull/34.
We will merge this test into that PR, and then that PR into eigenda-develop (our master branch).

### Note to reviewers

Please be very harsh on the golang test, as this same framework will be reused to add new tests for future features.

The "proper" way to implement this would probably have been to reuse op's [op-e2e framework](https://github.com/Layr-Labs/optimism/blob/8a1e07341d87afc923509920c2a54d1e9bd523f4/op-e2e/system/e2esys/setup.go#L349), but create a way to populate this system (equivalent to our harness in this PR) from a kurtosis backend. This would have taken me a lot more time to figure out however, and I feel like its something that the OP team might create at some point, so prefer to let them put in the work there and figure out how to do it properly, and we can potentially move to using that at a future point.

samlaf committed:
test(kt-devnet): add batcher failover test

samlaf committed:
chore(kt-devnet): use new proxy image v1.6.5 which has memstore rest routes

samlaf committed:
test(kt-devnet): fix TestFailover + add comments/logs

samlaf committed:
ci(kt-devnet): run new kurtosis failover test in ci

ethenotethan commented on kurtosis-devnet/tests/eigenda/failover_test.go:
```diff
@@ -0,0 +1,409 @@
+// We assume that this enclave is already running.
+const enclaveName = "eigenda-memstore-devnet"
+
+// TestFailover tests the failover behavior of the batcher, in response to the proxy returning 503 errors.
```
should we add a comment describing the parallelism limitations?

samlaf replied:
[8047e72](https://github.com/Layr-Labs/optimism/pull/33/commits/8047e7229c09dcb1f24374c45f3eaf882b5af7dd)

ethenotethan commented on kurtosis-devnet/tests/eigenda/failover_test.go:
```diff
@@ -0,0 +1,409 @@
+// The test then toggles the failover back off and checks that the batcher starts submitting EigenDA batches again.
+// The batches inbox transactions are queried via geth's GraphQL API.
+// TODO: We will also need to test the failover behavior of the node, which currently doesn't finalize after failover (fixed in https://github.com/Layr-Labs/optimism/pull/23)
+func TestFailover(t *testing.T) {
```
will we eventually wanna add tests ensuring failover to?
- keccak commitment mode
- 4844

samlaf replied:
keccak yes probably, I'll add a comment. 4844 I don't know... it's hard to implement and not a priority for me.

samlaf replied:
[bedc45f](https://github.com/Layr-Labs/optimism/pull/33/commits/bedc45f56bd36c8297815e34d9b969413997c2a0)

ethenotethan commented on kurtosis-devnet/tests/eigenda/failover_test.go:
```diff
@@ -0,0 +1,409 @@
+	// 1. Check that the original commitments are EigenDA
+	harness.requireBatcherTxsToBeFromLayer(t, sinceBlock, DALayerEigenDA)
+
+	// 2. Failover and check that the commitments are now EthDA
```
EthDA is just calldata here?

samlaf replied:
[bedc45f](https://github.com/Layr-Labs/optimism/pull/33/commits/bedc45f56bd36c8297815e34d9b969413997c2a0)

litt3 commented on kurtosis-devnet/tests/eigenda/failover_test.go:
```diff
@@ -0,0 +1,409 @@
+	})
+
+	// assume kurtosis is running and is at least at block 10 (just deploying the contracts takes more than 10 blocks)
+	require.GreaterOrEqual(t, harness.testStartL1BlockNum, uint64(10), "Test started too early in the chain")
```
magic number 10 is used in a lot of places - consider using a variable

samlaf replied:
Good callout. Hardcoded it and updated the graphql query functions to use both a FROM and a TO block. This then makes the semantics of the test much easier to understand, so I was able to document them more easily: [0dba0f2](https://github.com/Layr-Labs/optimism/pull/33/commits/0dba0f2d50b67455c0fca88213c154a6337a9972)

litt3 commented on kurtosis-devnet/tests/eigenda/failover_test.go:
```diff
@@ -0,0 +1,409 @@
+// Test Harness, which contains all the state needed to run the tests.
+// harness also defines some higher-level "require" methods that are used in the tests.
+type harness struct {
+	enclaveCtx          *enclaves.EnclaveContext
```
not currently used. is it planned for future use, or just there as a template for future tests that might need it?

samlaf replied:
Good call, added just in case, but agree its cleaner to remove. Done: [2277e3a](https://github.com/Layr-Labs/optimism/pull/33/commits/2277e3af0146623a473084841ea8845308211171)

litt3 commented on kurtosis-devnet/tests/eigenda/failover_test.go:
```diff
@@ -0,0 +1,409 @@
+		if batcherTx.daLayer != expectedLayer {
+			wrongCommitmentsToDiscard++
+		}
+		// as soon as we see 3 ethDA commitments, or an EigenDA commitment, we stop
```
doc seems to describe a previous version of the code which wasn't generalized to `expectedLayer`

samlaf replied:
[c63c4d0](https://github.com/Layr-Labs/optimism/pull/33/commits/c63c4d0075e303c4366026a9369054bcce283249)

litt3 commented on kurtosis-devnet/tests/eigenda/failover_test.go:
```diff
@@ -0,0 +1,409 @@
+		// Get the kurtosis tag
+		tag := field.Tag.Get("kurtosis")
+		if tag == "" {
+			continue // Skip fields without tags
```
What would be the reason for fields to exist without the tag? Is the idea that you'd add `kurtosis` fields to production structs?

samlaf replied:
None, good catch. This was claude generated. I updated to return error instead: [80fcebf](https://github.com/Layr-Labs/optimism/pull/33/commits/80fcebfdac5b01147181dd13547a68119b5b89b4)

litt3 commented on kurtosis-devnet/tests/eigenda/failover_test.go:
```diff
@@ -0,0 +1,409 @@
+	// NewServiceEndpoint   string `kurtosis:"new-service-name,port-name"`
+}
+
+func getEndpointsFromKurtosis(enclaveCtx *enclaves.EnclaveContext) (*EnclaveServiceEndpoints, error) {
```
A doc here would be helpful. It looks like this method replaces the service name in the `EnclaveServiceEndpoints` struct with a valid localhost endpoint?

litt3 commented on kurtosis-devnet/tests/eigenda/failover_test.go:
```diff
@@ -0,0 +1,409 @@
+// Update the proxy's memstore config to start returning 503 errors
+// Note: we have to GetConfig, update it and then UpdateConfig because the client doesn't implement a "patch" method,
+//
+//	even though the API does support it.
```
nit: weird formatting here, with newline and tab mid-sentence

samlaf replied:
[85b5c12](https://github.com/Layr-Labs/optimism/pull/33/commits/85b5c12b4bf96e8e8e062467a1e29df343ef5436)

samlaf committed:
test(kt-devnet): remove enclaveCtx from test harness

Was not being used and I had put it there only "in case"

samlaf committed:
test(kt-devnet): refactor failover test to not use hardcoded 10 constant

Define l1BlocksQueriedForBatcherTxs and better describe how tests use it
Also refactored graphql query to use both FROM and TO blocks, not only FROM. This makes the semantics of the tests easier to understand.

samlaf committed:
test(kt-devnet): better comments for requireBatcherTxsToBeFromLayer function

samlaf committed:
teest(kt-devnet): return err from getEndpointsFromKurtosis if EnclaveServiceEndpoints field doesn't have kurtosis tag

samlaf committed:
docs(kt-failover-test): rename getEndpointsFromKurtosis -> getPublicEndpointsFromKurtosis

Also better document the function to explain what its doing

samlaf committed:
style(kt-devnet): reformat a weirdly tabbed comment

samlaf committed:
docs(kt-failover-test): describe why we use graphql api insead of l1retriever api

samlaf committed:
docs(kt-failover-test): add note mentioning that kt tests need to be run sequentially

samlaf committed:
docs(kt-failover-test): update test name to TestFailoverToEthDACalldata

This is done to reflect the fact that batcher currently doesn't support failing over to calldata.

samlaf commented on kurtosis-devnet/tests/eigenda/failover_test.go:
```diff
@@ -0,0 +1,409 @@
+}
+
+// Fetches all the batch-inbox posted commitments from blockNum (inclusive) to current block.
+func fetchBatcherTxsSinceBlock(gethL1Endpoint string, batchInbox string, blockNum uint64) ([]BatcherTx, error) {
```
Added note in [6111c52](https://github.com/Layr-Labs/optimism/pull/33/commits/6111c52c354fa82f222b1e4f7e97d8da9dcd5a22)

Basically this feels easier for now. If we ever need to manage more complex state (like failing over to 4844 txs), then prob would be good to consider using L1Retriever instead.

samlaf committed:
style: fix lint

ethenotethan reviewed (approved):
LGTM 🛳️

litt3 commented on kurtosis-devnet/tests/eigenda/failover_test.go:
```diff
@@ -313,7 +341,7 @@ func getEndpointsFromKurtosis(enclaveCtx *enclaves.EnclaveContext) (*EnclaveServ
 		tag := field.Tag.Get("kurtosis")
 		if tag == "" {
-			continue // Skip fields without tags
+			return nil, fmt.Errorf("field %s doesn't have a kurtosis tag", field.Name)
```
If we require all tags to have `kurtosis` tag, can we just drop the tag requirement altogether?
